File: JavaProject/Reservations.java
import java.util.Date;
import java.util.List;
public class Reservations {
	private static Reservation[] all_reservations;
	private static List<Room> free_rooms;
	public static Reservation[] getAll_reservations() {
		return all_reservations;
	}

	public static void setAll_reservation(Reservation reservation, int index) {
		Reservations.all_reservations[index] = reservation;
	}

	public Reservation getOne_reservations(int index) {
		return all_reservations[index];
	}
	
	public static String Make_reservation(String client_id, Date start_date, Date end_date, int room_number)
	{
		int capacity = 0;
		if (Find_free_room(start_date,end_date,capacity).isEmpty())
		{
			return "nie ma wolnego pokoju";
		}
		else
		{
		Reservation res = new Reservation(client_id, start_date, end_date, room_number);
		
		setAll_reservation(res, getAll_reservations().length);
		return "rezerwacja dokonana";
		}
	}
	public void Cancel_reservation(String reservation_id)
	{
		
	}

	public static List<Room> Find_free_room(Date start_date, Date end_date, int capacity) {
		int number;
		for(int i=0; i<getAll_reservations().length;i++)
		{
			if(all_reservations[i].getStart_date().after(start_date) && all_reservations[i].getStart_date().equals(start_date))
			{
				number=all_reservations[i].getRoom_number();
				free_rooms.remove(number);

			}
		}
		
		return free_rooms;
	}

	public static List<Room> getFree_rooms() {
		return free_rooms;
	}

	public static void setFree_rooms(List<Room> free_rooms) {
		Reservations.free_rooms = free_rooms;
	}

	public static Room getFree_room(int index) {
		return free_rooms.get(index);
	}

}

File: CSharpProject/Hotel/Model/Reservations.cs
using Hotel.Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace HotelProject.Model
{
    public class Reservations:IReservations
    {
        List<Reservation> allReservations;

        public Reservations()
        {
            allReservations = new List<Reservation>();
            init();
        }
        public void init()
        { 
            DateTime data = new DateTime(2017, 01 , 12);
            DateTime datak = new DateTime(2017, 01, 14);
            Reservation przed = new Reservation("0", data , datak, "3",2);
            allReservations.Add(przed);

            data = new DateTime(2017, 01, 12);
            datak = new DateTime(2017, 01, 14);
            przed = new Reservation("1", data, datak, "2", 4);
            allReservations.Add(przed);

            data = new DateTime(2017, 01, 15);
            datak = new DateTime(2017, 01, 17);
            przed = new Reservation("2", data, datak, "1", 2);
            allReservations.Add(przed);


        }
        public void MakeReservation(String clientID, DateTime startDate, DateTime endDate, int roomNumber)
        {
            if (true) // tu ma być wywołana funkcja FindFreeRoom ale nie wiem jakie parametry
            {
                int idr=0;
                if (allReservations.Count!=0)
                idr = Int32.Parse(allReservations[allReservations.Count - 1].ReservationID) + 1;
                Reservation iten = new Reservation(idr.ToString(), startDate, endDate, clientID, roomNumber);
                allReservations.Add(iten);
            }
        }

        public void CancelReservation(String reservationID)
        {
            //reservationID = allReservations.Find(x => x.ReservationID.Contains(reservationID)).ReservationID; //trochę się rozpędziłem
            allReservations.RemoveAt(Int32.Parse(reservationID));
        }

        public RoomStatus CheckInClient(String reservationID, IRooms roomsCollection)
        {
            int i = 0;
            for (i = 0; i < allReservations.Count; i++)
                if (allReservations[i].ReservationID.Equals(reservationID))
                {
                    roomsCollection[allReservations[i].RoomNumber].Status = RoomStatus.Occupied;
                    break;
                }
            return roomsCollection[allReservations[i].RoomNumber].Status;
        }

        public void CheckOutClient(int roomNumber, IRooms roomsCollection)
        {
            roomsCollection.CheckOutClient(roomNumber);
        }

        public int[] FindFreeRoom(DateTime startDate, DateTime endDate, int capacity, Rooms roomsCollection)
        {
            if (allReservations.Count == 0)
                return new int[0];
            else
            {
                int index = 0;
                Room[] rooms = roomsCollection.ToArray();
                List<int> availableRooms = new List<int>(roomsCollection.ToRoomNumberArray());
                //int[] availableRooms = roomsCollection.ToRoomNumberArray();
                while (index < allReservations.Count)
                {
                    int roomNumber = allReservations[index].RoomNumber;
                    int i = 0;
                    for (i = 0; i < rooms.Length; i++)
                    {
                        if (rooms[i].RoomNumber == roomNumber)
                            break;
                    }
                    bool capacityIF = capacity == rooms[i].Capacity;
                    bool firstDateIF = (startDate < allReservations[index].StartDate) && (endDate <= allReservations[index].StartDate);
                    bool secondDateIF = (startDate >= allReservations[index].EndDate) && (endDate > allReservations[index].EndDate);

                    if (!(capacityIF && (firstDateIF || secondDateIF)))
                        availableRooms.Remove(roomNumber);
                    index++;
                }

                
                for (int j = 0; j < rooms.Length; j++)
                {
                    for (int i = 0; i < availableRooms.Count; i++)
                    {
                        if(availableRooms[i] == rooms[j].RoomNumber)
                        {
                            if (rooms[j].Capacity != capacity)
                                availableRooms.Remove(availableRooms[i]);
                            break;
                        }
                    }
                }

                return availableRooms.ToArray<int>();
            }
        }

        #region getters and setters
        public Reservation this[int i]
        {
            get { return allReservations[i]; }
            set { allReservations[i] = value; }
        }
        //IEnumerable<Reservation> AllReservations { get; }

        IEnumerable<Reservation> IReservations.AllReservations => throw new NotImplementedException();
        #endregion
    }
}
